A33-Fault-Injection.md

```diff
@@ -0,0 +1,262 @@
+Client-Side Fault Injection
+----
+* Author(s): lidiz
+* Approver: roth, ejona, dfawley
+* Status: In Review
+* Implemented in: all languages
+* Last updated: August 13th, 2020
+* Discussion at: https://groups.google.com/g/grpc-io/c/d9dI5Hy9zzk
+
+## Abstract
+
+Fault injection is an essential feature that assists developers in building fault-tolerant applications. To keep gRPC’s competitive advantage, this doc aims to design a fault injection mechanism on the client side for gRPC while allowing transparent transition from Envoy configuration.
+
+## Background
+
+### Envoy HTTPFilter
+
+Envoy has a network level filter named HTTP Connection Manager. It is responsible to handle common functionalities against HTTP connections and requests. In its `http_filters` field, users can set up multiple HTTPFilters. Each HTTPFilter requires at least a name. The name will be the identifier that assists future updates of filter configuration. For example, users can specify an `{"name": "envoy.cors"}` filter in `http_filters` with no content, then update the CORS filter’s configuration with the `typed_per_filter_config` field of a VirtualHost in RDS.
+
+Please refer to [A39: xDS HTTP Filters](https://github.com/grpc/proposal/pull/219) for how xDS HTTP Filters is processed in gRPC.
+
+### Fault Injection in Envoy xDS API
+
+Fault injection in xDS API is an HTTPFilter named `HTTPFault`. As mentioned above, the fault injection config can be provided by **LDS** (HTTP Connection Manager) and **RDS** (VirtualHost, Route, ClusterWeight). Besides, the fault injection filter offers [global Runtime settings](https://www.envoyproxy.io/docs/envoy/latest/configuration/http/http_filters/fault_filter#runtime) that can override global default fault injection filter config. For the explanation of Envoy Runtime, see section below.
+
+Currently, the Traffic Director (TD) isn’t forwarding HTTPFilter configs in HTTP Connection Manager from LDS, and there isn’t an exposed API allowing users to directly config fault injection for all HTTP connections. To pave gRPC’s way for more extensive control plane compatibility, we need to support fetching filter config from HTTP Connection Manager among LDS responses.
```

Comment:
Should we reference TD in the gRFC? Can we say certain control planes or certain xDS servers?

Reply:
Good catch, removed. After Mark's proposal, this restriction is outdated, there we will match our filter config accessing features with Envoy. So, there is no need to mention TD.